editor/plugins/node_3d_editor_plugin.cpp

```diff
@@ -6863,6 +6863,8 @@ void Node3DEditor::_sun_environ_settings_pressed() {
 	sun_environ_popup->set_position(pos - Vector2(sun_environ_popup->get_contents_minimum_size().width / 2, 0));
 	sun_environ_popup->reset_size();
 	sun_environ_popup->popup();
+	Rect2i safe_rect = DisplayServer::get_singleton()->screen_get_usable_rect();
+	DisplayServer::get_singleton()->window_set_popup_safe_rect(sun_environ_popup->get_window_id(), safe_rect);
```

Comment:
This will prevent closing the popup by clicking outside it. Probably X11 mouse handler should be changed to only trigger if mouse button state is changed (this need to be tested). On the other platforms, it should be the case already.

```diff
diff --git a/platform/linuxbsd/display_server_x11.cpp b/platform/linuxbsd/display_server_x11.cpp
index 07cb6a23e8..a2ec9cbe13 100644
--- a/platform/linuxbsd/display_server_x11.cpp
+++ b/platform/linuxbsd/display_server_x11.cpp
@@ -3250,7 +3250,7 @@ void DisplayServerX11::mouse_process_popups() {
 			XWindowAttributes root_attrs;
 			XGetWindowAttributes(x11_display, root, &root_attrs);
 			Vector2i pos = Vector2i(root_attrs.x + root_x, root_attrs.y + root_y);
-			if ((pos != last_mouse_monitor_pos) || (mask != last_mouse_monitor_mask)) {
+			if (mask != last_mouse_monitor_mask) {
 				if (((mask & Button1Mask) || (mask & Button2Mask) || (mask & Button3Mask) || (mask & Button4Mask) || (mask & Button5Mask))) {
 					List<WindowID>::Element *E = popup_list.back();
 					while (E) {
@@ -3272,7 +3272,6 @@ void DisplayServerX11::mouse_process_popups() {
 				}
 			}
 			last_mouse_monitor_mask = mask;
-			last_mouse_monitor_pos = pos;
 		}
 	}
 }
diff --git a/platform/linuxbsd/display_server_x11.h b/platform/linuxbsd/display_server_x11.h
index 63d32d939d..0a2194a23d 100644
--- a/platform/linuxbsd/display_server_x11.h
+++ b/platform/linuxbsd/display_server_x11.h
@@ -154,7 +154,6 @@ class DisplayServerX11 : public DisplayServer {
 	Map<WindowID, WindowData> windows;
 
 	unsigned int last_mouse_monitor_mask = 0;
-	Vector2i last_mouse_monitor_pos;
 	uint64_t time_since_popup = 0;
 
 	List<WindowID> popup_list;
```

Reply:
Thanks for the explanation. It would probably be best to split this change into a different PR, so that this PR can deal with the referenced issue.

Reply:
For the reference, "safe rect" is intended to prevent submenus from closing and re-opening when you click the parent menu item that was used to open the submenu.

Like the marked area here:
<img width="287" alt="Screenshot 2022-03-16 at 09 56 45" src="https://user-images.githubusercontent.com/7645683/158543284-d9faa25f-5dec-4fd5-a60a-a1d4ae2230e7.png">